Add option to disable reply keyboard

HappyTetrahedron opened the pull request:
Hi! First of all, thank you for this wonderful project, it's been a great asset to my 3D printing experience!

I've made a few small changes to accommodate my own preferences when it comes to Telegram Bot UX. I'm using these changes for myself, but I'd like to offer them to be integrated upstream, if you like them.

In this change, I'm adding a configuration option that disables the automatic reply keyboard that is sent together with the greeting message. I like the greeting - it lets me know immediately that everything is up and running - but the reply keyboard is, in my opinion, rather annoying. It takes away too much space on my screen, and I prefer using the commands menu or autocomplete.

I've made it so the default option is to show the keyboard (like before), but users who don't like it get the option to disable it.

HappyTetrahedron committed:
Add option to disable reply keyboard

aka13-404 commented:
We'll have to check that one out. Do I understand it right (currently not able to check) that this fully removes the pop-upable keyboard from the chat?

Also as in the other one, would you kindly target dev? Thx

HappyTetrahedron commented:
I've updated the target branch.

Yes, the intention is to fully remove the pop-upable keyboard from the chat. However, the "bot menu", which also lists all available commands, should stay and continue to be accessible.

I've realized that this might require some more work - in particular, if you've used the bot before with the keyboard enabled and then disable it later, it can still pop up if you scroll back far enough in the chat history. The old "show popup keyboard" chat events remain there forever. IIRC (and I'll have to double check this) there's a Telegram event to dismiss any existing popup keyboard in the chat; I might need to send such an event explicitly if the keyboard is disabled.

HappyTetrahedron committed:
[fix] Properly dismiss reply keyboard if it is not to be used

A Telegram reply keyboard sticks around until it is explicitly
dismissed. When the reply keyboard is disabled via config, we must send
a reply keyboard remove event, in case the keyboard was previously
enabled. Otherwise it will stick around even if disabled.

HappyTetrahedron commented:
Hello again! I kind of dropped the ball on this one - sorry about that! I've given this PR some love. Primarily, I've rebased it onto the current development branch.

I've also solved the problem I've mentioned in an earlier comment, in which the reply keyboard can sometimes stick around if it was first enabled and later disabled. The greeting message will now send an explicit "reply keyboard dismiss" event if the keyboard is disabled - this will clean up any stray reply keyboards that might be hanging around in the chat from earlier.

As I've mentioned, Telegram provides a bot menu in chats with bots, which lists all available bot commands (similar to what your reply keyboard does). This menu remains available for quick access to commands.

Here's what this looks like on Android:
![Screenshot_20260212-103109](https://github.com/user-attachments/assets/ee388a54-6842-4780-85b6-4245475ecd8c)
Chat without reply keyboard

![Screenshot_20260212-103127](https://github.com/user-attachments/assets/3289a89a-f94b-430a-8031-a54798e22def)
Same chat, with the "bot menu" expanded

Your bot is already configuring these using the `setMyCommands` Telegram API endpoint (cf. https://github.com/nlef/moonraker-telegram-bot/blob/master/bot/main.py#L1212 ) - which is how Telegram is aware of these commands and can display them. In my personal opinion, this makes the reply keyboard completely redundant, which is why I prefer to disable it.

Let me know if there's anything else I can do to improve this PR! I promise it won't take another year. ;)

nlef commented:
@HappyTetrahedron 
Sorry for the delay, I have a little time to work on the project and test the functionality. Everything works as advertised.
Thanks for the contribution.
